### ISSUE TITLE
Feature Request: Relief Valves ("lossy" backpressure)

### ISSUE THREAD

**Issue by Dryvnt** [alexpusch/pumps_rs#1]
I've been implementing a lot of bespoke pumps in a recent project, and this library finally provided me a metaphor for what we've been doing! I would love to be able to use this library for our use case, it feels like the perfect abstraction.

Simplifying our use case, we have a pipeline that maps type `A` to type `C` through `A -> B -> C`. It is vitally important that all `A`s are processed, whereas we only really care about the most recently produced `B`s, and would rather throw away old `B`s when the mapping to `C` can't keep up.

This desired "lossy" behavior of ours can be achieved by replacing the `mpsc` inside the pump with a `broadcast`. I am sure this is a lot easier to describe than actually implement.

I recognize that I could probably just implement this myself as a custom pump that I attach to the pipeline, but it would be nice to have good ergonomics. 

Example proposed API usage (you may probably think of something better):
```rust
let pipeline = pumps::Pipeline::from_stream(input)
    .map(a_to_b, Concurrency::serial()
        .backpressure(8)
        .relief_valve(ReliefValve::KeepNewest))
    .map(b_to_c, Concurrency::serial())
    .build();
```

Alternatively, to simplify implementation, relief valves could perhaps be added as a separate stage in the pipeline?

Example proposed API usage:
```rust
let pipeline = pumps::Pipeline::from_stream(input)
    .map(a_to_b, Concurrency::serial())
    .relief_valve(ReliefValve::keep_newest(8))
    .map(b_to_c, Concurrency::serial())
    .build();
```

---

If you're open to it, I would be willing to look at implementing this and making a PR over the weekend. Let me know if you think this would be a valuable addition, and which style of configuration you might prefer.

**Comment by alexpusch** [comment]
Hi @Dryvnt, glad to hear you find this crate useful! I've been thinking on similar feature myself. This reminds me of JS lodash [throttle](https://lodash.com/docs#throttle) and [debounce](https://lodash.com/docs#debounce). The way I think about this is that what I called `Concurrency` should actually be a more complex and customizable entity. Hopefully I'll work on this soon

**Comment by Dryvnt** [comment]
Great. I'll let you get around to it at your own pace, then. Let me know if you'd like any help or feedback.

Regarding the potential API of a "batteries included" relief valve: I've tried to map out our use case with this library, and I discovered an important detail that might inform the API design choices. I'll let you do with the information as you wish.

Our `A -> B` and `B -> C` stages cannot be represented as simple mapping functions. Every `A` can turn into multiple `B`, and it may take multiple `B` to create a single `C`. That's fine, I was expecting to have to create custom pumps for these anyway for a variety of reasons (e.g. we're calling a blocking C API using tokio `spawn_blocking`).

But it does mean that if relief valves are included as part of the `Concurrency` configuration, I'd have to do some extra identity mapping stage just to get that functionality. E.g.
```rust
let pipeline = Pipeline::from_stream(input)
    .pump(a_to_b_pump)
    .map(|x| futures::ready(x), Concurrency::serial().relief_valve(...))
    .pump(b_to_c_pump)
    .build();
```

Whether or not that's a problem is up to you.

**Comment by alexpusch** [comment]
If we're already talking about backpressure, I'd like to ask you something :) 
 
In the current API "backpressure" is configured on a pump output side - a pump accumulates unconsumed results
```rust
.map(foo, Concurrency::serial().backpressure(10)) // accumulates unconsumed results 
.map(bar, Concurrency::serial())
```
However, it is also common to regard backpressure on the *input* of operations. In other words, an operation can buffer input before processing

```rust
// should it actually be
.map(foo, Concurrency::serial())
.map(bar, Concurrency::serial().backpressure(10)) // buffers input before processing
```

Did the current placement of backpressure surprised you, or was it intuitive? I'm considering moving the backpressure config to the input side of pumps as I described

**Comment by raui100** [comment]
I haven't used the library yet but find it unintuitive to have the buffer "outside of the pump". 
This would mean that having a full buffer results from the following pump being slow. Having it "inside" means that a full buffer results from the corresponding pump being to slow which I find way more intuitive.

**Comment by alexpusch** [comment]
> I think either one can make intuitive sense, which may also be why neither seem to fit exactly. Zooming out and considering the options, might it work better to not have backpressure be part of the concurrency configuration?

This is a good point.  The reason I went with in-pump back pressure was my desire to use a mpsc::channel size parameter as natural backpressure. Since the pump creates the output channel, the backpressure configuration had to be passed to it. 
I now think this consideration isn't important enough and I think I'll change it. I'll remove backpressure from the concurrency config and add `.backpressure(n)` and `.backpressure_with_relief_valve(n)`

**Comment by alexpusch** [comment]
new methods available in v0.0.3
Thanks for the input @Dryvnt!
